xds/internal/xdsclient/channel.go

```diff
@@ -0,0 +1,288 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xdsclient
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/backoff"
+	igrpclog "google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/xds/bootstrap"
+	"google.golang.org/grpc/xds/internal/xdsclient/load"
+	"google.golang.org/grpc/xds/internal/xdsclient/transport"
+	"google.golang.org/grpc/xds/internal/xdsclient/transport/ads"
+	"google.golang.org/grpc/xds/internal/xdsclient/transport/lrs"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+)
+
+// channelEventHandler wraps callbacks used by the xdsChannel to notify the xDS
+// client about events on the channel. Methods in this interface may be invoked
+// concurrently and the xDS client implementation needs to handle them in a
+// thread-safe manner.
+type channelEventHandler interface {
+	// adsStreamFailure is called when the xdsChannel channel encounters an ADS
+	// stream failure.
+	adsStreamFailure(error)
+
+	// adsResourceUpdate is called when the xdsChannel receives an ADS response
+	// from the xDS management server. The callback is provided with the
+	// following:
+	//   - the resource type of the resources in the response
+	//   - a map of resources in the response, keyed by resource name
+	//   - the metadata associated with the response
+	//   - a callback to be invoked when the updated is processed
+	adsResourceUpdate(xdsresource.Type, map[string]ads.DataAndErrTuple, xdsresource.UpdateMetadata, func())
+
+	// adsResourceDoesNotExist is called when the xdsChannel determines that a
+	// requested ADS resource does not exist.
+	adsResourceDoesNotExist(xdsresource.Type, string)
+}
+
+// channelOptions holds the options for creating a new xdsChannel.
+type channelOptions struct {
```

Comment:
Optional: prepend this type and constructor with xds?

Reply:
I was little reluctant to change the ctor, because I don't like how the string looks with Go's initialism rules, i.e `newXDSChannel`, where the `XDS` now appears in all caps in the middle of the string. But I went ahead and did it anyways to be consistent with the rest of the file.

Reply:
I like xds as part of these typenames as this package is xdsclient so it encodes it there but these are used within same package so won't be xdsclient.Symbol it'll be symbol so I think it's better the symbol encodes xDS somewhere.